internal/pkg/util/fs/overlay/overlay_item_linux.go

```diff
@@ -0,0 +1,297 @@
+		return fmt.Errorf("internal error: unrecognized image type in overlay.Item.Mount() (type: %v)", i.Type)
+	}
+}
+
+// mountDir mounts directory-based OverlayItems. This involves bind-mounting
+// followed by remounting of the directory onto itself. This pattern of
+// bind-mount followed by remount allows application of more restrictive mount
+// flags than are in force on the underlying filesystem.
+func (i *Item) mountDir() error {
+	var err error
+	if len(i.DirToMount) < 1 {
+		i.DirToMount = i.BarePath
+	}
+
+	if err = EnsureOverlayDir(i.DirToMount, false, 0); err != nil {
+		return fmt.Errorf("error accessing directory %s: %s", i.DirToMount, err)
+	}
+
+	sylog.Debugf("Performing identity bind-mount of %q", i.DirToMount)
+	if err = syscall.Mount(i.DirToMount, i.DirToMount, "", syscall.MS_BIND, ""); err != nil {
+		return fmt.Errorf("failed to bind %s: %s", i.DirToMount, err)
+	}
+
+	// Best effort to cleanup mount
+	defer func() {
+		if err != nil {
+			sylog.Debugf("Encountered error with current OverlaySet; attempting to unmount %q", i.DirToMount)
+			syscall.Unmount(i.DirToMount, syscall.MNT_DETACH)
+		}
+	}()
+
+	// Try to perform remount
+	sylog.Debugf("Performing remount of %q", i.DirToMount)
+	if err = syscall.Mount("", i.DirToMount, "", syscall.MS_REMOUNT|syscall.MS_BIND, ""); err != nil {
+		return fmt.Errorf("failed to remount %s: %s", i.DirToMount, err)
+	}
+
+	return nil
+}
+
+// mountWithFuse mounts an image to a temporary directory using a specified fuse
+// tool. It also verifies that fusermount is present before performing the
+// mount.
+func (i *Item) mountWithFuse(fuseMountTool string) error {
+	var err error
+	fuseMountCmd, err := bin.FindBin(fuseMountTool)
+	if err != nil {
+		return fmt.Errorf("use of image %q as overlay requires %s to be installed: %s", i.BarePath, fuseMountTool, err)
+	}
+
+	// Even though fusermount is not needed for this step, we shouldn't perform
+	// the mount unless we have the necessary tools to eventually unmount it
+	_, err = bin.FindBin("fusermount")
+	if err != nil {
+		return fmt.Errorf("use of image %q as overlay requires fusermount to be installed: %s", i.BarePath, err)
+	}
+
+	// Obtain parent directory in which to create overlay-related mount
+	// directories. See https://github.com/apptainer/singularity/pull/5575 for
+	// related discussion.
+	parentDir, err := i.GetSecureParentDir()
+	if err != nil {
+		return fmt.Errorf("error while trying to create parent dir for overlay %q: %s", i.BarePath, err)
+	}
+	fuseMountDir, err := os.MkdirTemp(parentDir, "overlay-mountpoint-")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary dir %q for overlay %q: %s", fuseMountDir, i.BarePath, err)
+	}
+
+	// Best effort to cleanup temporary dir
+	defer func() {
+		if err != nil {
+			sylog.Debugf("Encountered error with current OverlaySet; attempting to remove %q", fuseMountDir)
+			os.Remove(fuseMountDir)
+		}
+	}()
+
+	execCmd := exec.Command(fuseMountCmd, i.BarePath, fuseMountDir)
```

Comment:
The FUSE mount of an image will need to be able to mount the image read-only, when the overlay request is read-only.

For squashfs there's no issue... it can only be read only.

For extfs, only one mount can safely be performed in read-write mode. If we want to use a read-only overlay in extfs format across multiple concurrent runs of `singularity`, the image mount must be `ro` as well as the `overlayfs` portion.